common/src/math.rs

```diff
@@ -3,97 +3,463 @@
-        na::Vector4::new(self.0.x, self.0.y, self.0.z, w)
+        let w = (sqr(gans.x) + sqr(gans.y) + sqr(gans.z) + na::one()).sqrt();
+        MVector(na::Vector4::new(gans.x, gans.y, gans.z, w))
+    }
+    #[inline]
+    pub fn zero() -> Self
+    {
+        Self(na::zero())
+    }
+    pub fn origin() -> Self
+    {
+        Self(na::Vector4::new(na::zero(),na::zero(),na::zero(),na::one()))
+    }
+    #[inline]
+    pub fn normalize(self) -> Self
+    {
+        Self(self.0.normalize())
+    }
+    #[inline]
+    pub fn x() -> Self
+    {
+        Self(na::Vector4::x())
+    }
+    #[inline]
+    pub fn y() -> Self
+    {
+        Self(na::Vector4::y())
+    }
+    #[inline]
+    pub fn z() -> Self
+    {
+        Self(na::Vector4::z())
+    }
+    #[inline]
+    pub fn w() -> Self
+    {
+        Self(na::Vector4::w())
+    }
+    #[inline]
+    pub fn new(x: N, y: N, z: N, w: N) -> Self
+    {
+        MVector(na::Vector4::new(x,y,z,w))
+    }
+    #[inline]
+    pub fn xyz(self) -> na::Vector3<N>
+    {
+        self.0.xyz()
     }
 }
 
-/// Point or plane reflection around point or normal `p`
-pub fn reflect<N: RealField + Copy>(p: &na::Vector4<N>) -> na::Matrix4<N> {
-    na::Matrix4::<N>::identity()
-        - minkowski_outer_product(p, p) * na::convert::<_, N>(2.0) / mip(p, p)
+impl<N: RealField> Mul<MIsometry<N>> for MIsometry<N>
+{
+    type Output = MIsometry<N>;
+    #[inline]
+    fn mul(self, rhs: MIsometry<N>) -> Self::Output
+    {
+        MIsometry(self.0 * rhs.0)
+    }
+}
+
+impl<N: RealField> Mul<N> for MVector<N>
+{
+    type Output = MVector<N>;
+    #[inline]
+    fn mul(self, rhs: N) -> Self::Output
+    {
+        MVector(self.0 * rhs)
+    }
+}
+
+impl<N: RealField> Div<N> for MVector<N>
+{
+    type Output = MVector<N>;
+    #[inline]
+    fn div(self, rhs: N) -> Self::Output
+    {
+        MVector(self.0 / rhs)
+    }
+}
+
+impl<N: RealField> Add for MVector<N>
+{
+    type Output = Self;
+    #[inline]
+    fn add(self, other: Self) -> Self
+    {
+        Self(self.0 + other.0)
+    }
+}
+
+impl<N: RealField> Sub for MVector<N>
+{
+    type Output = Self;
+    #[inline]
+    fn sub(self, other: Self) -> Self
+    {
+        Self(self.0 - other.0)
+    }
+}
+
+impl<N: RealField> Neg for MVector<N>
+{
+    type Output = Self;
+    #[inline]
+    fn neg(self) -> Self
+    {
+        Self(-self.0)
+    }
+}
+
+impl<N: RealField> Mul<MVector<N>> for MIsometry<N>
+{
+    type Output = MVector<N>;
+    #[inline]
+    fn mul(self, rhs: MVector<N>) -> Self::Output
+    {
+        MVector(self.0 * rhs.0)
+    }
+}
+
+impl<N: RealField> Mul<N> for MIsometry<N>
+{
+    type Output = MIsometry<N>;
+    #[inline]
+    fn mul(self, rhs: N) -> Self::Output
+    {
+        MIsometry(self.0 * rhs)
+    }
+}
+
+impl<N: RealField + Copy> std::ops::AddAssign for MVector<N>
+{
+    #[inline]
+    fn add_assign(&mut self, other: Self)
+    {
+        *self = *self + other;
```

Comment:
I wonder if `self.0 += other.0` would be better here.

Reply:
back when i tried that, looks like it just calls add_assign and just recurses, so i guess for brevity i could add a macro or something, though im not familiar with macros

Reply:
It won't recurse, because `self.0` is a different type than `self`.

Reply:
oh misread, that makes sense